src/Stream/GetProductsStreamer.php

```diff
@@ -0,0 +1,295 @@
+            return $metadata[self::METADATA_ITEMS_PER_PAGE];
+        }
+
+        return abs($metadata[self::METADATA_ITEMS_PER_PAGE] * $metadata[self::METADATA_PAGE] - $metadata[self::METADATA_TOTALS_COUNT]);
+    }
+
+    /**
+     * @param $streamedFileName
+     * @return array|null
+     */
+    public function getProductsIds($streamedFileName)
+    {
+        $callbackResult = [];
+
+        $stream = $this->openStream($streamedFileName);
+
+        $productIdListener = new RegexListener([
+            '/entries/*/id' => function ($productId) use (&$callbackResult) {
+                $callbackResult[] = $productId;
+            }
+        ]);
+
+        $parser = new Parser($stream, $productIdListener);
+        $parser->parse();
+
+        $this->closeStream($stream);
+
+        return $callbackResult;
+    }
+
+    /**
+     * @param $streamedFileName
+     * @param array $validItemIds
+     * @param array $variations
+     * @param $createProduct
+     * @return array
+     */
+    public function processProducts($streamedFileName, $validItemIds, $variations, callable $createProduct)
+    {
+        $productsCount = $this->getProductsCount($streamedFileName);
+
+        if ($productsCount <= 0) {
+            return [];
+        }
+
+        $trackSkippedProductsIds = [];
+
+        for ($i = 0; $i < $productsCount; $i++) {
+            $stream = $this->openStream($streamedFileName);
+
+            $getProductListener = new RegexListener();
+            $parser = new Parser($stream, $getProductListener);
+            $getProductListener->setMatch([
+                '/entries/' . $i  => function ($productData) use ($createProduct, $variations, $validItemIds, &$trackSkippedProductsIds, $parser) {
+                    if (!in_array($productData['id'], $validItemIds)) {
+                        return;
+                    }
+
+                    /** @var Product $product */
+                    $product = $createProduct($productData);
+
+                    while (($variation = array_shift($variations[$product->getItemId()]))) {
+                        $continueProcess = $product->processVariation($variation);
+
+                        if (!$continueProcess) {
+                            continue;
+                        }
+
+                        if (isset($variation['itemImages'])) {
+                            $product->processImages($variation['itemImages']);
+                        }
+
+                        if (isset($variation['variationProperties'])) {
+                            $product->processVariationsProperties($variation['variationProperties']);
+                        }
+
+                        if (isset($variation['properties'])) {
+                            $product->processVariationSpecificProperties($variation['properties']);
+                        }
+                    }
+
+                    if ($product->hasValidData()) {
+                        $this->wrapper->wrapItem($product->getResults());
```

Comment:
First we need to ensure to write the correct data e.g. id, name, etc.
Since we want to write each piece of data after the other, not the whole product at once.